encodings/fastlanes/src/bitpacking/compute/search_sorted.rs

```diff
@@ -29,69 +86,92 @@ fn search_sorted_typed<T>(
+        + AsPrimitive<u64>,
+{
+    let native_value: T = value.cast(array.dtype())?.try_into()?;
+    search_sorted_native(array, native_value, side)
+}
+
+/// Native variant of search_sorted that operates over Rust unsigned integer types.
+fn search_sorted_native<T>(
+    array: &BitPackedArray,
+    value: T,
+    side: SearchSortedSide,
+) -> VortexResult<SearchResult>
+where
+    T: NativePType + BitPacking + AsPrimitive<usize> + AsPrimitive<u64>,
 {
-    let unwrapped_value: T = value.cast(array.dtype())?.try_into()?;
     if let Some(patches_array) = array.patches() {
         // If patches exist they must be the last elements in the array, if the value we're looking for is greater than
         // max packed value just search the patches
-        if unwrapped_value.as_() > array.max_packed_value() {
-            search_sorted(&patches_array, value.clone(), side)
+        let usize_value: usize = value.as_();
+        if usize_value > array.max_packed_value() {
+            search_sorted_u64(&patches_array, value.as_(), side)
         } else {
-            Ok(BitPackedSearch::new(array).search_sorted(&unwrapped_value, side))
+            Ok(BitPackedSearch::<'_, T>::new(array).search_sorted(&value, side))
         }
     } else {
-        Ok(BitPackedSearch::new(array).search_sorted(&unwrapped_value, side))
+        Ok(BitPackedSearch::<'_, T>::new(array).search_sorted(&value, side))
     }
 }
 
 /// This wrapper exists, so that you can't invoke SearchSorted::search_sorted directly on BitPackedArray as it omits searching patches
 #[derive(Debug)]
-struct BitPackedSearch {
-    packed: PrimitiveArray,
+struct BitPackedSearch<'a, T> {
+    // NOTE: caching this here is important for performance, as each call to `maybe_null_slice`
+    //  invokes a call to DType <> PType conversion
+    packed_maybe_null_slice: &'a [T],
     offset: usize,
     length: usize,
     bit_width: usize,
-    min_patch_offset: Option<usize>,
-    validity: Validity,
+    first_invalid_idx: usize,
 }
 
-impl BitPackedSearch {
-    pub fn new(array: &BitPackedArray) -> Self {
+impl<'a, T: BitPacking + NativePType> BitPackedSearch<'a, T> {
+    pub fn new(array: &'a BitPackedArray) -> Self {
+        let min_patch_offset = array
+            .patches()
+            .and_then(|p| {
+                SparseArray::try_from(p)
+                    .vortex_expect("Only sparse patches are supported")
+                    .min_index()
+            })
+            .unwrap_or_else(|| array.len());
+        let first_null_idx = match array.validity() {
+            Validity::NonNullable | Validity::AllValid => array.len(),
+            Validity::AllInvalid => 0,
+            Validity::Array(varray) => {
+                // In sorted order, nulls come after all the non-null values.
+                varray.with_dyn(|a| a.as_bool_array_unchecked().true_count())
+            }
+        };
+
+        let first_invalid_idx = cmp::min(min_patch_offset, first_null_idx);
```

Comment:
Oh right, yes the patches will overlap with nulls and we just need the min.

Reply:
yea, lets us do 1 compare instead of 2 in the inner loop

Reply:
doubt it matters too much but might as well